Create Kernel Tuner Cache Python Library

warnio opened the pull request.

warnio committed:
Move read/write function to a different location

warnio committed:
Rename file

warnio committed:
Remove KTLibrary folder

warnio committed:
Ignore docstring requirement in tests

warnio committed:
Make code conform with Ruff project config

tdejong00 reviewed (commented):
De Ruff-gerelateerde aanpassingen zijn prima 👍

rmgerritse commented:
Yes, de aanpassing en renaming zijn goed!

rmgerritse reviewed (commented):
Alles ziet er goed uit !

TobiasHofstra commented on test/test_cachefile_read_write.py:
```diff
@@ -1,51 +1,36 @@
 
-    p = tmp_output_file
-    
     # cache_file = read_cache_file(file_path)
```
Deze line moet weg toch? En de comments een paar regels verder op over hashes zijn toch niet meer relevant?

warnio committed:
Create global variable for sample cache file

warnio committed:
Test reading open cache files

warnio committed:
Fix cache file reading

warnio committed:
Rename variable

warnio committed:
Remove newline at EOF

warnio committed:
Add close cache file functionality

warnio committed:
Add functionality for opening a cache file

warnio committed:
Rename function

warnio committed:
Implement appending cache lines to cache files

warnio committed:
Ensure open or close cache files in read_cache_file via keyword args

warnio committed:
Simplify calls marked with parametrize

warnio committed:
Improve function names and documentation

warnio committed:
Create separate functions for ensuring open and closed cache files

warnio committed:
Assert read_cache_file does not change the file being read by default

warnio committed:
Don't allow ensure_open and ensure_closed to be simultaneously set

warnio committed:
Simplify cache file function names

warnio committed:
Fix closing improperly opened cache files

warnio committed:
Add comments

warnio committed:
Cover additional case where cache is invalid

warnio committed:
Add custom error type

warnio committed:
Rename tests to be more descriptive

warnio committed:
Cleanup unneeded file

warnio committed:
Invoke super

warnio committed:
Fix ErrorConfig bug

Remove its str superclass, so that default is invoked in JSONDecoder. It turns out that inheriting the string superclass made ErrorConfig falsey, on which code in common.py depended. Now that code has also changed.

warnio committed:
Add tests and encode branch for empty list

warnio committed:
Add support for encoding non-json objects

tdejong00 commented on kernel_tuner/cache/file.py:
```diff
@@ -1,47 +1,174 @@
+
+def close_cache(filename: PathLike):
+    """Opens any cache file. Needs to read the whole file in order to guarantee this."""
+    read_cache(filename, ensure_closed=True)
```
Volgens mij staan deze twee docstrings verkeerd om

tdejong00 reviewed (approved):
De aanpassingen zien er goed uit, heb naast mijn enkele comment geen opmerkingen

rmgerritse reviewed (approved):
Ziet er goed uit, niks op aan te merken

warnio committed:
Test encode cache last

warnio committed:
Add test for appending cache lines to empty cache

warnio committed:
Simplify reading and writing cache files

warnio committed:
Restructure code and simplify append_cache_line function signature

warnio committed:
Add important documentation to function append_cache_line

TobiasHofstra commented:
Ziet er goed uit, mooie test structuur voor de cachefile encoder!

warnio committed:
Add cache creation functionality

warnio committed:
Implement reading existing cache files

Note: validation is yet to be added

warnio committed:
Improve test

warnio committed:
Add extra test for reading and writing cache files

warnio committed:
Return native Python objects when reading cache

warnio committed:
Implement appending cache lines and fix bug

warnio committed:
Add filepath property to Cache

warnio committed:
Reserve a subset of tune_param_keys

warnio committed:
Raise error when tunable param has invalid value

warnio committed:
Add test cases for getting a default from Cache.Lines

warnio committed:
Remove print statement

warnio committed:
Add type verification for appending lines

warnio committed:
Remove frozendict and use of tuple in properties

warnio committed:
Improve readability

warnio committed:
Rename CacheLinePosition

warnio committed:
Slightly alter type signature to only accept lists

warnio committed:
Add JSON validation

warnio committed:
Extract modules for readability

warnio committed:
Remove automatic conversion of cache files in cache library

warnio committed:
Remove comments

TobiasHofstra reviewed (commented):
Ik vind de aanpassingen die gemaakt zijn goed!

Misschien kan het handig zijn om te documenteren welke geïmplementeerde functies in ``cache.py`` de al bestaande functies in ``util.py`` dienen te vervangen, en hoe een nieuwe ``Cache`` object gemaakt kan worden. Dat is vooral heel nuttig voor de klant.

warnio committed:
Remove unused import

warnio committed:
Improve documentation

warnio commented:
@TobiasHofstra 

> Misschien kan het handig zijn om te documenteren welke geïmplementeerde functies in `cache.py` de al bestaande functies in `util.py` dienen te vervangen, en hoe een nieuwe `Cache` object gemaakt kan worden. Dat is vooral heel nuttig voor de klant.

Ik heb niet gedocumenteerd hoe ze bestaande functies dienen te vervangen, maar ik heb nu wel uitgebreid gedocumenteerd hoe je de classes in `cache.py` kan gebruiken en cache files daarmee kan veranderen.

warnio committed:
Improve code readability

warnio committed:
Return json serializable objects from __getitem__

warnio committed:
Replace printf by print

warnio committed:
Make validation more strict

warnio committed:
Improve documentation and naming

tdejong00 reviewed (approved):
Het ziet er erg netjes en goed gedocumenteerd uit! Ik ben naast de documentatie foutjes, die je ter plekke hebt opgelost, niks geks tegengekomen. Wat mij betreft kan dit gemerged worden :+1:
